File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/repository/heap/FieldValueEntity.java
package com.github.ol_loginov.heaplibweb.repository.heap;

import com.github.ol_loginov.heaplibweb.repository.EntityInstance;
import jakarta.persistence.Entity;
import jakarta.persistence.Id;
import jakarta.persistence.IdClass;
import jakarta.persistence.Table;
import lombok.*;

import java.io.Serializable;

@Entity
@Table(name = "FieldValue")
@IdClass(FieldValueEntity.PK.class)
@Getter
@Setter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@AllArgsConstructor
public class FieldValueEntity implements EntityInstance {
	@Data
	@AllArgsConstructor
	@NoArgsConstructor
	public static class PK implements Serializable {
		private long javaClassId;
		private long definingInstanceId;
		private int fieldId;
	}

	@Id
	private long javaClassId;
	@Id
	private long definingInstanceId;
	@Id
	private int fieldId;
	private boolean staticFlag;
	private String value;
	private Long valueInstanceId;
}

File: heap-web/src/main/java/com/github/ol_loginov/heaplibweb/repository/heap/FieldValueEntity.kt
package com.github.ol_loginov.heaplibweb.repository.heap

import com.github.ol_loginov.heaplibweb.hprof.HprofValueType
import com.github.ol_loginov.heaplibweb.repository.EntityInstance

class FieldValueEntity(
    var instanceId: Long,
    var fieldId: Int,
    var value: String,
    var valueInstanceId: Long = 0
) : EntityInstance {
    internal constructor() : this(0, 0, "", 0)

    companion object {
        fun anyToValueText(type: HprofValueType, any: Any) = when (type) {
            HprofValueType.Object -> (any as ULong).toString()
            HprofValueType.Char -> (any as Char).code.toString()
            else -> any.toString()
        }
    }

    override fun toString(): String = "FieldValueEntity(instanceId=$instanceId,fieldId=$fieldId,value=$value,valueInstanceId=$valueInstanceId)"
}